### ISSUE TITLE
Handle case of when scopes for Developer Key change, but users already have token

### ISSUE THREAD

**Issue by ssciolla** [tl-its-umich-edu/canvas-course-manager-next#104]
As the project evolves, we may need to change the scopes associated with the API Developer Key used to create in the OAuth workflow (implemented in pending PR #92). However, the tokens we've stored in the database for users will still be associated with an older set of scopes. How can we implement a change management strategy such that if the scopes change, tokens are invalidated and/or the OAuth flow is re-initiated?

Idea(s):
- Add management process to remove all Canvas tokens when scopes change
- Re-initiate the OAuth flow for a user if they encounter an error related to scopes not matching.

**Comment by pushyamig** [comment]
This is the last issue remaining on OAuth process. We can either bulk delete token manually during deployment when scopes do changes. But if the scopes were not selected by mistake during the configuration ( by mistake) then we will endup deleting all the scope if want to handle it programmatically. 

I am fine with manual approach for  first release. We can discuss on this

**Comment by pushyamig** [comment]
Testing is needed.

**Comment by pushyamig** [comment]
As Prep before, 
1. I removed one of the API like `'url:GET|/api/v1/accounts', from developer console api scopes and canvas.scope.ts.
2. I deleted the DB that was created before
3. Did the LTI launch, Launch worked, token generated.
After step
1. Enabled  `'url:GET|/api/v1/accounts',` API from the developer Key console
2. added these API’s tp canvas.scopes.ts file.
3. Restarted the server

Results: LTI launch happened, when accessing that specific API in merge section, I saw the authorization page came up and new token is generated. So things worked as expected. 

I consider this as Pass. @ssciolla

**Comment by ssciolla** [comment]
Question about the first three steps: So when you remove a scope, but don't touch the database or existing tokens, what happens? Does the old token still work?

**Comment by pushyamig** [comment]
Resolved.

**Comment by ssciolla** [comment]
I'll QA this.

**Comment by ssciolla** [comment]
I think this issue is resolved by the handling of the `InvalidTokenInterceptor`.

Here is how I tested.

Scenario 1:
1) Removed the PUT `courses` scope from both the `canvas.scopes.ts` file and the API Developer Key.
2) Tried changing the course name with the old token still in the database
3) Observed error and re-authorization flow was triggered. (This doesn't help the user fix this problem, since we're not granting this scope anymore through the OAuth, but we also shouldn't be leaving API calls that require the scope we removed).

Scenario 2 (picking up after S1):
1) After re-authorizing and getting a new token without the scope granted, I added back the scope in the configuration and in Canvas.
2) I tried to re-name the course, observed the error, and followed the re-authorization flow.
3) After re-authorizing, I was able to successfully change the course name.

So, basically, as scopes are added or subtracted, if a scope is used but was not part of a token's initial grant or not included in the allowed scopes, the process will delete the token and ask the user to re-authorize. This should serve us well for change management. If a new scope is added for a new feature, users will need to re-authorize when they start using the feature. If a scope is removed along with a feature, the user may never notice because they won't be using that scope (at least through the UI).

I'm moving this to Done.
